airflow/jobs/local_task_job.py

```diff
@@ -95,6 +95,15 @@ def signal_handler(signum, frame):
                 # Monitor the task to see if it's done
                 return_code = self.task_runner.return_code()
                 if return_code is not None:
+                    print(self.task_instance.state)
+                    # there is one case we should not treat non zero return
+                    # code as failed: the job has been killed externally.
+                    if return_code != 0 and ((not self.terminating) or
+                                             self.task_instance.state ==
```

Comment:
Additionally though: why do we care about calling on-failure/raising when the task_instance is already set to failure.

You are overloading two subtly different concepts here:

1. The task executed to completion, but it was marked as a failure.
2. The task process died without managing to tidy up.

Case 1 is already handled and the TI ends up in a "terminal" state successfully. It's Case 2 is the only place we need to change. What do you think?

Reply:
The task state is usually set in another process, therefore the final state is not visible to the local task job. We need to make sure we get the correct state before moving forward.

Reply:
good point. I need to call refresh from db before proceed.

Reply:
> Additionally though: why do we care about calling on-failure/raising when the task_instance is already set to failure.
> 
> You are overloading two subtly different concepts here:
> 
> 1. The task executed to completion, but it was marked as a failure.
> 2. The task process died without managing to tidy up.
> 
> Case 1 is already handled and the TI ends up in a "terminal" state successfully. It's Case 2 is the only place we need to change. What do you think?

In theory there is only one case, the process exited with non zero exit code. We need to set the correct states for task and job.
1. The task state may or may not be set correctly, that's what on_failure is handling.
2. The job state is definitely not set as FAILED, so that's why raise another exception to let BaseJob to handle this.

Reply:
TBH JobState isn't used anyway (so long as it's not still running).

It's open to interpretation when a job should be marked as failed. If the task ran to completion then the job did what it was told to. Is that a failure? What about when the task is in "up_for_retry" -- what state should the Job be in?

My first though is that the job should be marked as success if the task is in any state other than "running" at the end, and only failure if the Task didn't update it's state.

Reply:
> TBH JobState isn't used anyway (so long as it's not still running).
> 
> It's open to interpretation when a job should be marked as failed. If the task ran to completion then the job did what it was told to. Is that a failure? What about when the task is in "up_for_retry" -- what state should the Job be in?
> 
> My first though is that the job should be marked as success if the task is in any state other than "running" at the end, and only failure if the Task didn't update it's state.

Hmm, I see your point but I think we should better to be consistent about interpreting the meaning of Job.state across or job types?

e.g, for BackfillJob, if one of the underlying tasks fail, it will be marked as failure as well
on backfill_job.py:L792
"""
  err = self._collect_errors(ti_status=ti_status, session=session)
                if err:
                    raise AirflowException(err)
""" 

So I think it might be better to have job state reflects the state of underlying task instance